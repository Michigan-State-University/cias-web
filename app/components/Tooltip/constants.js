export const TooltipType = {
  WARNING: 'WARNING',
};
