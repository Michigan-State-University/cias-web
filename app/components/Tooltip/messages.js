import { defineMessages } from 'react-intl';

export const scope = 'app.containers.Tooltip';

export default defineMessages({
  warningTitle: {
    id: `${scope}.warningTitle`,
    defaultMessage: 'Important note',
  },
});
