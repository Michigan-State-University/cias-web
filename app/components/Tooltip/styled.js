import styled from 'styled-components';
import ReactTooltip from 'react-tooltip';

import { boxShadows } from 'theme';

export const StyledTooltip = styled(ReactTooltip).attrs({
  suppressClassNameWarning: true,
})`
  box-shadow: ${boxShadows.selago};
  border-radius: 5px;
  max-width: 380px;
  padding: 15px;
  opacity: 1 !important;
`;
