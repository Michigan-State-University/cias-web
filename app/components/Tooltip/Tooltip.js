/**
 *
 * Tooltip
 *
 */

import React, { memo, useCallback, useEffect, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import ReactTooltip from 'react-tooltip';

import useKeyPress from 'utils/useKeyPress';
import { KeyCodes } from 'utils/constants';
import useOutsideClick from 'utils/useOutsideClick';

import { TOOLTIP_PORTAL_ID } from 'containers/App/constants';

import Box from 'components/Box';
import Img from 'components/Img';
import Text from 'components/Text/Text'; // Import like that due to dependency-cycle
import Portal from 'components/Portal';

import { StyledTooltip } from './styled';

/**
 * WARNING: It should be placed after the component with `data-tip` property, not before
 * otherwise it will not render properly
 */
const Tooltip = ({
  id,
  children,
  visible,
  text,
  icon,
  content,
  place,
  stretchContent,
  backgroundColor,
  ...restProps
}) => {
  const tooltipRef = useRef();
  const contentRef = useRef();
  const [isHovered, setIsHovered] = useState(false);

  const onFocusIn = () => setIsHovered(true);
  const onFocusOut = () => setIsHovered(false);

  const showTooltip = () => ReactTooltip.show(tooltipRef.current);
  const hideTooltip = () => ReactTooltip.hide(tooltipRef.current);

  const shouldShowTooltip = visible && isHovered;

  useKeyPress(KeyCodes.ESC, onFocusOut, shouldShowTooltip);
  useOutsideClick([tooltipRef, contentRef], onFocusOut, shouldShowTooltip);

  useEffect(() => {
    if (shouldShowTooltip) showTooltip();
    else hideTooltip();

    return () => {
      hideTooltip();
    };
  }, [shouldShowTooltip]);

  const onTooltipClick = (event) => {
    const portal = document.getElementById(TOOLTIP_PORTAL_ID);

    if (portal?.contains(event.target)) {
      event.preventDefault();
      event.stopPropagation();
    }
  };

  const getContent = (dataTip) => (
    <Box ref={contentRef}>
      <Text>{dataTip || text}</Text>
      {content}
    </Box>
  );

  const calculateTooltipPosition = useCallback(
    ({ left, top }, event, triggerElement, tooltipElement) => ({
      top,
      left: typeof tooltipElement === 'string' ? left : Math.max(left, 0),
    }),
    [],
  );

  const stretchContentStyle = stretchContent
    ? { width: '100%', height: '100%' }
    : {};

  return (
    <Box
      display="flex"
      {...stretchContentStyle}
      {...restProps}
      onClick={onTooltipClick}
    >
      <Box
        ref={tooltipRef}
        data-tip=""
        data-for={id}
        onMouseEnter={onFocusIn}
        onTouchStart={onFocusIn}
        {...stretchContentStyle}
      >
        {icon && <Img src={icon} alt="?" />}
        {children && <div style={stretchContentStyle}>{children}</div>}
      </Box>

      {shouldShowTooltip && (
        <Portal id={TOOLTIP_PORTAL_ID}>
          <StyledTooltip
            visible={visible}
            id={id}
            type="light"
            effect="solid"
            multiline
            getContent={getContent}
            delayHide={200}
            afterHide={onFocusOut}
            overridePosition={place ? undefined : calculateTooltipPosition}
            place={place}
            backgroundColor={backgroundColor}
          />
        </Portal>
      )}
    </Box>
  );
};

Tooltip.propTypes = {
  id: PropTypes.string.isRequired,
  children: PropTypes.node,
  icon: PropTypes.any,
  text: PropTypes.string,
  content: PropTypes.node,
  visible: PropTypes.bool,
  place: PropTypes.oneOf(['top', 'right', 'bottom', 'left']),
  stretchContent: PropTypes.bool,
  backgroundColor: PropTypes.string,
};

Tooltip.defaultProps = {
  visible: true,
};

export default memo(Tooltip);
