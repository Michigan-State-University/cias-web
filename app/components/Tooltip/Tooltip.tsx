/**
 *
 * Tooltip
 *
 */

import React, {
  memo,
  useCallback,
  useEffect,
  useRef,
  useState,
  ReactElement,
} from 'react';
import ReactTooltip from 'react-tooltip';

import useKeyPress from 'utils/useKeyPress';
import { KeyCodes } from 'utils/constants';
import useOutsideClick from 'utils/useOutsideClick';

import { TOOLTIP_PORTAL_ID } from 'containers/App/constants';

import Box from 'components/Box';
import Img from 'components/Img';
import Text from 'components/Text/Text'; // Import like that due to dependency-cycle
import Portal from 'components/Portal';

import { StyledTooltip } from './styled';

type TooltipProps = {
  id: string;
  children?: ReactElement;
  visible?: boolean;
  text?: string;
  icon?: any;
  content?: ReactElement;
  place?: 'top' | 'right' | 'bottom' | 'left';
  stretchContent?: boolean;
  backgroundColor?: string;
} & Record<string, unknown>;

/**
 * WARNING: It should be placed after the component with `data-tip` property, not before
 * otherwise it will not render properly
 */
const Tooltip = ({
  id,
  children,
  visible = true,
  text,
  icon,
  content,
  place,
  stretchContent,
  backgroundColor,
  ...restProps
}: TooltipProps) => {
  const tooltipRef = useRef<HTMLDivElement>();
  const contentRef = useRef<HTMLDivElement>();
  const [isHovered, setIsHovered] = useState(false);

  const onFocusIn = () => setIsHovered(true);
  const onFocusOut = () => setIsHovered(false);

  const showTooltip = () =>
    tooltipRef.current && ReactTooltip.show(tooltipRef.current);
  const hideTooltip = () => ReactTooltip.hide(tooltipRef.current);

  const shouldShowTooltip = visible && isHovered;

  useKeyPress(KeyCodes.ESC, onFocusOut, shouldShowTooltip);
  useOutsideClick([tooltipRef, contentRef], onFocusOut, shouldShowTooltip);

  useEffect(() => {
    if (shouldShowTooltip) showTooltip();
    else hideTooltip();

    return () => {
      hideTooltip();
    };
  }, [shouldShowTooltip]);

  const onTooltipClick = (event: MouseEvent) => {
    const portal = document.getElementById(TOOLTIP_PORTAL_ID);

    if (portal?.contains(event.target as Node)) {
      event.preventDefault();
      event.stopPropagation();
    }
  };

  const getContent = (dataTip: string) => (
    <Box ref={contentRef}>
      <Text>{dataTip || text}</Text>
      {content}
    </Box>
  );

  const calculateTooltipPosition = useCallback(
    ({ left, top }, event, triggerElement, tooltipElement) => ({
      top,
      left: typeof tooltipElement === 'string' ? left : Math.max(left, 0),
    }),
    [],
  );

  const stretchContentStyle = stretchContent
    ? { width: '100%', height: '100%' }
    : {};

  return (
    <Box
      display="flex"
      {...stretchContentStyle}
      {...restProps}
      onClick={onTooltipClick}
    >
      <Box
        ref={tooltipRef}
        data-tip=""
        data-for={id}
        onMouseEnter={onFocusIn}
        onTouchStart={onFocusIn}
        {...stretchContentStyle}
      >
        {icon && <Img src={icon} alt="?" />}
        {children && <div style={stretchContentStyle}>{children}</div>}
      </Box>

      {shouldShowTooltip && (
        <Portal id={TOOLTIP_PORTAL_ID}>
          <StyledTooltip
            visible={visible}
            id={id}
            type="light"
            effect="solid"
            multiline
            getContent={getContent}
            delayHide={200}
            afterHide={onFocusOut}
            overridePosition={place ? undefined : calculateTooltipPosition}
            place={place}
            backgroundColor={backgroundColor}
          />
        </Portal>
      )}
    </Box>
  );
};

export default memo(Tooltip);
