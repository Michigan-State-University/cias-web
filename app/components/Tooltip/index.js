/**
 *
 * Tooltip
 *
 */

import React from 'react';
import PropTypes from 'prop-types';

import Box from 'components/Box';
import Img from 'components/Img';
import Text from 'components/Text';

import { StyledTooltip } from './styled';

const Tooltip = ({ id, children, text, icon, ...restProps }) => (
  <Box display="flex" {...restProps}>
    {icon && <Img src={icon} alt="?" data-tip data-for={id} />}
    {children && (
      <div data-tip data-for={id}>
        {children}
      </div>
    )}
    <StyledTooltip id={id} type="light" effect="solid" multiline>
      <Text>{text}</Text>
    </StyledTooltip>
  </Box>
);

Tooltip.propTypes = {
  id: PropTypes.string.isRequired,
  children: PropTypes.node,
  icon: PropTypes.any,
  text: PropTypes.string,
};

export default Tooltip;
