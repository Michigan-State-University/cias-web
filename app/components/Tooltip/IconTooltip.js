import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import { useIntl } from 'react-intl';

import Column from 'components/Column';
import Row from 'components/Row';
import { WarningIcon } from 'components/Icons';
import Text from 'components/Text';

import Tooltip from './Tooltip';
import { TooltipType } from './constants';
import messages from './messages';

export const IconTooltip = ({
  content,
  text,
  children,
  type,
  ...tooltipProps
}) => {
  const { formatMessage } = useIntl();

  const { icon, title } = useMemo(() => {
    switch (type) {
      case TooltipType.WARNING:
      default:
        return {
          icon: <WarningIcon />,
          title: formatMessage(messages.warningTitle),
        };
    }
  });

  return (
    <Tooltip
      {...tooltipProps}
      content={
        <Row>
          <Column width="min-content" mr={8}>
            {icon}
          </Column>

          <Column>
            <Text fontWeight="bold" mb={10}>
              {title}
            </Text>

            {text ?? content}
          </Column>
        </Row>
      }
    >
      {children}
    </Tooltip>
  );
};

IconTooltip.propTypes = {
  content: PropTypes.node,
  text: PropTypes.string,
  children: PropTypes.node,
  type: PropTypes.string,
};
