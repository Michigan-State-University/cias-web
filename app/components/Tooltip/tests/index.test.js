/**
 *
 * Tests for Tooltip
 *
 */

import React from 'react';
import { render } from '@testing-library/react';
import 'jest-styled-components';

import Tooltip from '../index';

const defaultProps = {
  id: 'assda1-123dsa',
};

describe('<Tooltip />', () => {
  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(<Tooltip {...defaultProps}>Tooltip text</Tooltip>);
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot', () => {
    const { container } = render(
      <Tooltip {...defaultProps}>Tooltip text</Tooltip>,
    );
    expect(container).toMatchSnapshot();
  });
});
