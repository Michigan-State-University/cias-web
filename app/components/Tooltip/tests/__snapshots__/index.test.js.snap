// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tooltip /> Should render and match the snapshot 1`] = `
.c0 {
  width: auto;
  height: auto;
  display: block;
  border-radius: 5px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 {
  font-family: DM Sans;
  font-size: 13px;
  line-height: 130%;
}

<div>
  <div
    class="c0"
    display="flex"
  >
    <div
      currentitem="false"
      data-for="assda1-123dsa"
      data-tip="true"
    >
      Tooltip text
    </div>
    <div
      class="__react_component_tooltip t00000000-0000-0000-0000-000000000000 place-top type-dark"
      data-id="tooltip"
      id="assda1-123dsa"
    >
      <style>
        
  	.t00000000-0000-0000-0000-000000000000 {
	    color: #fff;
	    background: #222;
	    border: 1px solid transparent;
  	}

  	.t00000000-0000-0000-0000-000000000000.place-top {
        margin-top: -10px;
    }
    .t00000000-0000-0000-0000-000000000000.place-top::before {
        border-top: 8px solid transparent;
    }
    .t00000000-0000-0000-0000-000000000000.place-top::after {
        border-left: 8px solid transparent;
        border-right: 8px solid transparent;
        bottom: -6px;
        left: 50%;
        margin-left: -8px;
        border-top-color: #222;
        border-top-style: solid;
        border-top-width: 6px;
    }

    .t00000000-0000-0000-0000-000000000000.place-bottom {
        margin-top: 10px;
    }
    .t00000000-0000-0000-0000-000000000000.place-bottom::before {
        border-bottom: 8px solid transparent;
    }
    .t00000000-0000-0000-0000-000000000000.place-bottom::after {
        border-left: 8px solid transparent;
        border-right: 8px solid transparent;
        top: -6px;
        left: 50%;
        margin-left: -8px;
        border-bottom-color: #222;
        border-bottom-style: solid;
        border-bottom-width: 6px;
    }

    .t00000000-0000-0000-0000-000000000000.place-left {
        margin-left: -10px;
    }
    .t00000000-0000-0000-0000-000000000000.place-left::before {
        border-left: 8px solid transparent;
    }
    .t00000000-0000-0000-0000-000000000000.place-left::after {
        border-top: 5px solid transparent;
        border-bottom: 5px solid transparent;
        right: -6px;
        top: 50%;
        margin-top: -4px;
        border-left-color: #222;
        border-left-style: solid;
        border-left-width: 6px;
    }

    .t00000000-0000-0000-0000-000000000000.place-right {
        margin-left: 10px;
    }
    .t00000000-0000-0000-0000-000000000000.place-right::before {
        border-right: 8px solid transparent;
    }
    .t00000000-0000-0000-0000-000000000000.place-right::after {
        border-top: 5px solid transparent;
        border-bottom: 5px solid transparent;
        left: -6px;
        top: 50%;
        margin-top: -4px;
        border-right-color: #222;
        border-right-style: solid;
        border-right-width: 6px;
    }
  
      </style>
      <p
        class="c1"
      />
    </div>
  </div>
</div>
`;
